x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/type/InvalidMappedField.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.esql.type;
+
+import org.elasticsearch.xpack.ql.type.EsField;
+
+import java.util.Map;
+import java.util.Set;
+import java.util.TreeMap;
+
+/**
+ * Representation of field mapped differently across indices.
+ * Used during mapping discovery only.
+ * Note that, for backwards compatibility purposes, this class extends the QL class InvalidMappedField,
+ * and uses the same simple name and constructor signature. This allows serialization of query plans to
+ * work across upgrades involving older servers that use the QL version of this class.
+ * The additional field <code>typesToIndices</code> is not serialized in any version of this class,
+ * because that information is not required through the cluster, only surviving as long as the Analyser
+ * phase of query planning.
+ * TODO: Once QL and ESQL code bases are split, we should incorporate the error message from the QL version into this class.
+ * Breaking the inheritance requires that we port IndexResolver to ESQL, so it cannot be done yet.
+ */
+public class InvalidMappedField extends org.elasticsearch.xpack.ql.type.InvalidMappedField {
```

Comment:
This scares me a bit TBH...
Eg. we still use QL IndexResolver that returns a QL `InvalidMappedField`. 
Could it for instance fail de/serialization, since `PlanNamedTypes` only knows the ES|QL version of `InvalidMappedField`?

Reply:
InvalidMappedField does not need to survive serialisation, and the changes are purely additive. So in my opinion, either approach works as well:

* Edit InvalidMappedField directly and add a comment about the additions being purely for ESQL use cases
* Extend InvalidMappedField with the ESQL version, so QL code is left untouched, and this is a step towards the QL/ESQL split.

In both cases, the common behaviour is identically serializable, and the information lost when serializating the ESQL version to QL is of no consequence. I can add a comment about that.

Reply:
We took the latter approach, since it is a step towards splitting from QL. We did this for InvalidMappedField, but not for FieldAttribute, since that was much, much messier to port, and the changes were seen as an improvement anyway.